File: src/catalog.ts
/**
 * Read and safely mutate the episode catalog (`seasons.md`).
 *
 * Operates on the raw markdown lines so surrounding prose is never disturbed —
 * only per-season table rows change. Fence-aware, so the illustrative example
 * table in seasons.md is ignored.
 */

import { readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

export const DEFAULT_PATH = join(dirname(fileURLToPath(import.meta.url)), "..", "seasons.md");

const SEASON_HEADING = /^##\s+Season\s+(\d+)/i;
const ACTIVE_SEASON = /Active season:\s*\*{0,2}(\d+)/i;
const PLACEHOLDER_MARK = "no episodes yet";
const EMPTY = "—";

export class CatalogError extends Error {}

export interface Row {
  season: number;
  ep: number | null; // null for the placeholder row
  album: string;
  artist: string;
  host: string;
  status: string;
  dir: string;
  published: string;
  lineno: number; // 0-based index into the file's line list
}

export interface AssignResult {
  season: number;
  episode: number;
  dir: string;
  action: "claimed" | "appended";
}

const pad2 = (n: number): string => String(n).padStart(2, "0");

export function slug(text: string): string {
  const s = text.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  return s || "untitled";
}

function cells(line: string): string[] {
  let parts = line.trim().split("|").map((c) => c.trim());
  if (parts.length && parts[0] === "") parts = parts.slice(1);
  if (parts.length && parts[parts.length - 1] === "") parts = parts.slice(0, -1);
  return parts;
}

const isTableLine = (line: string): boolean => line.trim().startsWith("|");
const isSeparator = (line: string): boolean => isTableLine(line) && /^[|:\-\s]+$/.test(line.trim());
const isHeader = (line: string): boolean => cells(line)[0] === "Ep";
const isPlaceholder = (line: string): boolean => line.toLowerCase().includes(PLACEHOLDER_MARK);

/** Blank out lines inside ``` / ~~~ code fences, preserving line indices. */
function maskFences(lines: string[]): string[] {
  const out: string[] = [];
  let inFence = false;
  for (const line of lines) {
    const stripped = line.replace(/^\s+/, "");
    if (stripped.startsWith("```") || stripped.startsWith("~~~")) {
      inFence = !inFence;
      out.push("");
      continue;
    }
    out.push(inFence ? "" : line);
  }
  return out;
}

interface TableLoc {
  headerIdx: number;
  dataIndices: number[];
  hasPlaceholder: boolean;
}

function locateTable(lines: string[], season: number): TableLoc {
  const masked = maskFences(lines);
  let start = -1;
  for (let i = 0; i < masked.length; i++) {
    const m = masked[i]!.match(SEASON_HEADING);
    if (m && parseInt(m[1]!, 10) === season) {
      start = i;
      break;
    }
  }
  if (start === -1) throw new CatalogError(`Season ${season} not found in catalog`);

  let headerIdx = -1;
  for (let i = start + 1; i < masked.length; i++) {
    if (SEASON_HEADING.test(masked[i]!)) break;
    if (isTableLine(masked[i]!) && isHeader(masked[i]!)) {
      headerIdx = i;
      break;
    }
  }
  if (headerIdx === -1) throw new CatalogError(`Season ${season} has no table`);

  const dataIndices: number[] = [];
  let hasPlaceholder = false;
  let i = headerIdx + 1;
  while (i < masked.length && isTableLine(masked[i]!)) {
    if (!isSeparator(masked[i]!) && !isHeader(masked[i]!)) {
      dataIndices.push(i);
      if (isPlaceholder(masked[i]!)) hasPlaceholder = true;
    }
    i++;
  }
  return { headerIdx, dataIndices, hasPlaceholder };
}

export function read(path: string = DEFAULT_PATH): string {
  return readFileSync(path, "utf-8");
}

export function activeSeason(text: string): number {
  const m = text.match(ACTIVE_SEASON);
  if (!m) throw new CatalogError("no 'Active season:' marker found in catalog");
  return parseInt(m[1]!, 10);
}

export function rowsForSeason(text: string, season: number): Row[] {
  const lines = text.split("\n");
  const { dataIndices } = locateTable(lines, season);
  const out: Row[] = [];
  for (const idx of dataIndices) {
    if (isPlaceholder(lines[idx]!)) continue;
    const c = cells(lines[idx]!);
    while (c.length < 7) c.push("");
    const ep = /^\d+$/.test(c[0]!) ? parseInt(c[0]!, 10) : null;
    out.push({
      season, ep, album: c[1]!, artist: c[2]!, host: c[3]!,
      status: c[4]!, dir: c[5]!, published: c[6]!, lineno: idx,
    });
  }
  return out;
}

export function nextEpisode(text: string, season: number): number {
  const eps = rowsForSeason(text, season)
    .map((r) => r.ep)
    .filter((e): e is number => e !== null);
  return eps.length ? Math.max(...eps) + 1 : 1;
}

function formatRow(ep: number, album: string, artist: string, host: string,
                   status: string, dir: string, published: string): string {
  return `| ${pad2(ep)} | ${album} | ${artist} | ${host} | ${status} | ${dir} | ${published} |`;
}

/** Claim a matching `planned` row or append the next episode; set it in-production. */
export function assign(album: string, artist: string, host: string,
                       season?: number, path: string = DEFAULT_PATH): AssignResult {
  const text = readFileSync(path, "utf-8");
  const seas = season ?? activeSeason(text);
  const lines = text.split("\n");
  const { headerIdx, dataIndices, hasPlaceholder } = locateTable(lines, seas);

  for (const r of rowsForSeason(text, seas)) {
    if (r.status.toLowerCase() === "planned"
        && r.album.toLowerCase() === album.toLowerCase()
        && r.artist.toLowerCase() === artist.toLowerCase()) {
      const ep = r.ep ?? nextEpisode(text, seas);
      const d = `S${pad2(seas)}E${pad2(ep)}-${slug(album)}`;
      lines[r.lineno] = formatRow(ep, r.album, r.artist, host, "in-production", d, EMPTY);
      writeFileSync(path, lines.join("\n") + "\n", "utf-8");
      return { season: seas, episode: ep, dir: d, action: "claimed" };
    }
  }

  const ep = nextEpisode(text, seas);
  const d = `S${pad2(seas)}E${pad2(ep)}-${slug(album)}`;
  const newRow = formatRow(ep, album, artist, host, "in-production", d, EMPTY);
  if (hasPlaceholder) {
    const ph = dataIndices.find((i) => isPlaceholder(lines[i]!))!;
    lines[ph] = newRow;
  } else {
    const insertAt = (dataIndices.length ? dataIndices[dataIndices.length - 1]! : headerIdx) + 1;
    lines.splice(insertAt, 0, newRow);
  }
  writeFileSync(path, lines.join("\n") + "\n", "utf-8");
  return { season: seas, episode: ep, dir: d, action: "appended" };
}

export function setStatus(season: number, ep: number, status: string,
                          published?: string, path: string = DEFAULT_PATH): void {
  const text = readFileSync(path, "utf-8");
  const lines = text.split("\n");
  for (const r of rowsForSeason(text, season)) {
    if (r.ep === ep) {
      const pub = published ?? r.published;
      lines[r.lineno] = formatRow(ep, r.album, r.artist, r.host, status, r.dir, pub);
      writeFileSync(path, lines.join("\n") + "\n", "utf-8");
      return;
    }
  }
  throw new CatalogError(`S${pad2(season)}E${pad2(ep)} not found in catalog`);
}

File: pipeline/catalog.py
"""Read and safely mutate the episode catalog (``seasons.md``).

The catalog is the authoritative source for episode numbers and status. These
helpers operate on the raw markdown lines so the surrounding prose, legend, and
"How the Producer Agent uses this file" section are never disturbed — only the
per-season table rows change.

Numbering rule (from seasons.md): episode = highest Ep in the season + 1, per
season (numbers reset each season). Season planning = pre-added ``planned`` rows,
which ``assign`` claims instead of appending a duplicate.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

DEFAULT_PATH = Path(__file__).resolve().parent.parent / "seasons.md"

SEASON_HEADING = re.compile(r"^##\s+Season\s+(\d+)", re.I)
ACTIVE_SEASON = re.compile(r"Active season:\s*\*{0,2}(\d+)", re.I)
COLUMNS = ["Ep", "Album", "Artist", "Host", "Status", "Dir", "Published"]
PLACEHOLDER_MARK = "no episodes yet"
EMPTY = "—"


class CatalogError(Exception):
    pass


@dataclass
class Row:
    season: int
    ep: int | None       # None for the placeholder row
    album: str
    artist: str
    host: str
    status: str
    dir: str
    published: str
    lineno: int          # 0-based index into the file's line list


def slug(text: str) -> str:
    s = re.sub(r"[^a-z0-9]+", "-", text.lower()).strip("-")
    return s or "untitled"


def _cells(line: str) -> list[str]:
    parts = [c.strip() for c in line.strip().split("|")]
    # split of "| a | b |" yields ['', 'a', 'b', ''] -> drop the outer empties
    if parts and parts[0] == "":
        parts = parts[1:]
    if parts and parts[-1] == "":
        parts = parts[:-1]
    return parts


def _is_table_line(line: str) -> bool:
    return line.strip().startswith("|")


def _is_separator(line: str) -> bool:
    return _is_table_line(line) and set(line.strip()) <= set("|-: ")


def _is_header(line: str) -> bool:
    return _cells(line)[:1] == ["Ep"]


def _is_placeholder(line: str) -> bool:
    return PLACEHOLDER_MARK in line.lower()


def _mask_fences(lines: list[str]) -> list[str]:
    """Blank out lines inside ``` / ~~~ code fences, preserving line indices.

    seasons.md carries an *illustrative* example table inside a fence (heading and
    all). Masking it keeps the parser from mistaking the example for real data,
    while index-preservation means mutations still target the right real lines.
    """
    out, in_fence = [], False
    for line in lines:
        stripped = line.lstrip()
        if stripped.startswith("```") or stripped.startswith("~~~"):
            in_fence = not in_fence
            out.append("")
            continue
        out.append("" if in_fence else line)
    return out


# --- reading -----------------------------------------------------------------

def read(path: str | Path = DEFAULT_PATH) -> str:
    return Path(path).read_text(encoding="utf-8")


def active_season(text: str) -> int:
    m = ACTIVE_SEASON.search(text)
    if not m:
        raise CatalogError("no 'Active season:' marker found in catalog")
    return int(m.group(1))


def _locate_table(lines: list[str], season: int):
    """Return (header_idx, data_indices, has_placeholder) for a season's table.

    Detection runs over a fence-masked view so the illustrative example is
    ignored; the returned indices are valid against the original ``lines``.
    """
    masked = _mask_fences(lines)
    start = None
    for i, line in enumerate(masked):
        m = SEASON_HEADING.match(line)
        if m and int(m.group(1)) == season:
            start = i
            break
    if start is None:
        raise CatalogError(f"Season {season} not found in catalog")

    header_idx = None
    for i in range(start + 1, len(masked)):
        if SEASON_HEADING.match(masked[i]):
            break
        if _is_table_line(masked[i]) and _is_header(masked[i]):
            header_idx = i
            break
    if header_idx is None:
        raise CatalogError(f"Season {season} has no table")

    data_indices, has_placeholder = [], False
    i = header_idx + 1
    while i < len(masked) and _is_table_line(masked[i]):
        if not _is_separator(masked[i]) and not _is_header(masked[i]):
            data_indices.append(i)
            if _is_placeholder(masked[i]):
                has_placeholder = True
        i += 1
    return header_idx, data_indices, has_placeholder


def rows_for_season(text: str, season: int) -> list[Row]:
    lines = text.splitlines()
    _, data_indices, _ = _locate_table(lines, season)
    out = []
    for idx in data_indices:
        if _is_placeholder(lines[idx]):
            continue
        c = _cells(lines[idx])
        c += [""] * (len(COLUMNS) - len(c))  # pad short rows
        ep = int(c[0]) if c[0].isdigit() else None
        out.append(Row(season, ep, c[1], c[2], c[3], c[4], c[5], c[6], idx))
    return out


def next_episode(text: str, season: int) -> int:
    eps = [r.ep for r in rows_for_season(text, season) if r.ep is not None]
    return (max(eps) + 1) if eps else 1


def _format_row(ep: int, album: str, artist: str, host: str,
                status: str, dir_: str, published: str) -> str:
    return f"| {ep:02d} | {album} | {artist} | {host} | {status} | {dir_} | {published} |"


# --- mutation ----------------------------------------------------------------

def assign(album: str, artist: str, host: str, season: int | None = None,
           path: str | Path = DEFAULT_PATH) -> dict:
    """Claim a matching ``planned`` row or append the next episode.

    Sets the row's status to ``in-production`` and fills its Dir. Returns
    ``{season, episode, dir, action}`` where action is 'claimed' or 'appended'.
    """
    path = Path(path)
    text = path.read_text(encoding="utf-8")
    season = season if season is not None else active_season(text)
    lines = text.splitlines()
    header_idx, data_indices, has_placeholder = _locate_table(lines, season)

    # Look for a matching planned row to claim.
    for r in rows_for_season(text, season):
        if (r.status.lower() == "planned"
                and r.album.lower() == album.lower()
                and r.artist.lower() == artist.lower()):
            ep = r.ep if r.ep is not None else next_episode(text, season)
            d = f"S{season:02d}E{ep:02d}-{slug(album)}"
            lines[r.lineno] = _format_row(ep, r.album, r.artist, host,
                                          "in-production", d, EMPTY)
            path.write_text("\n".join(lines) + "\n", encoding="utf-8")
            return {"season": season, "episode": ep, "dir": d, "action": "claimed"}

    # Otherwise append the next episode.
    ep = next_episode(text, season)
    d = f"S{season:02d}E{ep:02d}-{slug(album)}"
    new_row = _format_row(ep, album, artist, host, "in-production", d, EMPTY)
    if has_placeholder:
        ph = next(i for i in data_indices if _is_placeholder(lines[i]))
        lines[ph] = new_row
    else:
        insert_at = (data_indices[-1] if data_indices else header_idx + 1) + 1
        lines.insert(insert_at, new_row)
    path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    return {"season": season, "episode": ep, "dir": d, "action": "appended"}


def set_status(season: int, ep: int, status: str, published: str | None = None,
               path: str | Path = DEFAULT_PATH) -> None:
    path = Path(path)
    text = path.read_text(encoding="utf-8")
    lines = text.splitlines()
    for r in rows_for_season(text, season):
        if r.ep == ep:
            pub = published if published is not None else r.published
            lines[r.lineno] = _format_row(ep, r.album, r.artist, r.host,
                                          status, r.dir, pub)
            path.write_text("\n".join(lines) + "\n", encoding="utf-8")
            return
    raise CatalogError(f"S{season:02d}E{ep:02d} not found in catalog")


# --- cli ---------------------------------------------------------------------

def run_cli(args) -> int:
    path = Path(args.file) if args.file else DEFAULT_PATH
    text = read(path)
    if args.catalog_cmd == "next":
        season = args.season if args.season is not None else active_season(text)
        print(next_episode(text, season))
    elif args.catalog_cmd == "list":
        season = args.season if args.season is not None else active_season(text)
        rows = rows_for_season(text, season)
        print(f"Season {season} (active={active_season(text)}): {len(rows)} episode(s)")
        for r in rows:
            print(f"  E{r.ep:02d}  {r.status:<14} {r.album} — {r.artist} "
                  f"(host {r.host})  dir={r.dir}  pub={r.published}")
    elif args.catalog_cmd == "assign":
        result = assign(args.album, args.artist, args.host, args.season, path)
        print(f"{result['action']}: S{result['season']:02d}E{result['episode']:02d} "
              f"→ {result['dir']}")
    return 0
